runtimes/core/src/api/gateway/websocket.rs

```diff
@@ -0,0 +1,329 @@
+
+        for protocol in header_value.to_str()?.split(',') {
+            let protocol = protocol.trim();
+            if protocol.starts_with(ENCORE_DEV_HEADERS) {
+                let data = protocol.strip_prefix(ENCORE_DEV_HEADERS).unwrap();
+                let decoded = URL_SAFE_NO_PAD.decode(data)?;
+                let auth_data: AuthHeaderMap = serde_json::from_slice(&decoded)?;
+
+                for (name, value) in auth_data.0 {
+                    if is_forbidden_request_header(&name) {
+                        return Err(anyhow!("header {name} not allowed to be set"));
+                    }
+                    upstream_request.append_header(name, value)?;
+                }
+            } else {
+                filterd_protocols.push(protocol);
+            }
+        }
+
+        if !filterd_protocols.is_empty() {
+            upstream_request.append_header(SEC_WEBSOCKET_PROTOCOL, filterd_protocols.join(", "))?;
+        }
+    }
+
+    Ok(())
+}
+
+struct AuthHeaderMap(HashMap<HeaderName, HeaderValue>);
+
+impl<'de> Deserialize<'de> for AuthHeaderMap {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        deserializer.deserialize_map(AuthHeaderMapVisitor)
+    }
+}
+
+struct AuthHeaderMapVisitor;
+
+impl<'de> Visitor<'de> for AuthHeaderMapVisitor {
+    type Value = AuthHeaderMap;
+
+    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+        formatter.write_str("AuthHeaderMap")
+    }
+
+    fn visit_map<A>(self, mut access: A) -> Result<Self::Value, A::Error>
+    where
+        A: serde::de::MapAccess<'de>,
+    {
+        let mut map = HashMap::with_capacity(access.size_hint().unwrap_or_default());
+
+        while let Some((key, value)) = access.next_entry::<&str, &str>()? {
+            let name = HeaderName::from_str(key).map_err(de::Error::custom)?;
+            let value = HeaderValue::from_str(value).map_err(de::Error::custom)?;
+
+            map.insert(name, value);
+        }
+
+        Ok(AuthHeaderMap(map))
+    }
+}
+
+// see https://developer.mozilla.org/en-US/docs/Glossary/Forbidden_header_name
+fn is_forbidden_request_header(name: &HeaderName) -> bool {
+    use http::header::*;
+    match *name {
+        ACCEPT_CHARSET
+        | ACCEPT_ENCODING
+        | ACCESS_CONTROL_REQUEST_HEADERS
+        | ACCESS_CONTROL_REQUEST_METHOD
+        | CONNECTION
+        | CONTENT_LENGTH
+        | COOKIE
+        | DATE
+        | DNT
+        | EXPECT
+        | HOST
+        | ORIGIN
+        | REFERER
+        | TE
+        | TRAILER
+        | TRANSFER_ENCODING
+        | UPGRADE
+        | VIA => true,
+        ref n if n == HeaderName::from_static("keep-alive") => true,
+        ref n if n == HeaderName::from_static("permissions-policy") => true,
+        ref n => {
+            let name = n.to_string();
+            name.strip_prefix("sec-").is_some() || name.strip_prefix("proxy-").is_some()
```

Comment:
Is there not an `has_prefix` or similar?

Reply:
there is a `starts_with`! fixed